File: apps/chaja-app/src/ipc/merge.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

export type MergeStrategy =
  | "fast-forward-only"
  | "fast-forward-or-merge"
  | "no-fast-forward";

export type MergeResult =
  | { kind: "already-up-to-date" }
  | { kind: "fast-forward"; new_head: string }
  | { kind: "merged"; new_head: string }
  | { kind: "conflict"; paths: string[] };

export function mergeBranch(
  repoPath: string,
  source: string,
  strategy: MergeStrategy,
): Promise<MergeResult> {
  return invoke<MergeResult>("merge_branch", { repoPath, source, strategy });
}

export function abortMerge(repoPath: string): Promise<void> {
  return invoke<void>("abort_merge", { repoPath });
}

File: crates/yryvu-bridge/src/commands/merge.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::PathBuf;

use crate::backend::{GitBackend, MergeResult, MergeStrategy};
use crate::repo::GixBackend;

#[tauri::command]
pub async fn merge_branch(
    repo_path: String,
    source: String,
    strategy: MergeStrategy,
) -> Result<MergeResult, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .merge_branch(&PathBuf::from(&repo_path), &source, strategy)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}
